plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +241,185 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+        } else {
+            nuageVspDevice = nuageVspDevices.iterator().next();
+            nuageVspHost = _hostDao.findById(nuageVspDevice.getHostId());
+            _hostDao.loadDetails(nuageVspHost);
+        }
+
+        Map<String, String> paramsTobeUpdated = new HashMap<String, String>();
+        if (!Strings.isNullOrEmpty(command.getHostName()) &&
+                !command.getHostName().equals(nuageVspHost.getDetails().get("hostname"))) {
+            paramsTobeUpdated.put("name", "Nuage VSD - " + command.getHostName());
+            paramsTobeUpdated.put("hostname", command.getHostName());
+        }
+
+        if (!Strings.isNullOrEmpty(command.getUserName()) &&
+                !command.getUserName().equals(nuageVspHost.getDetails().get("cmsuser"))) {
+            paramsTobeUpdated.put("cmsuser", command.getUserName());
+        }
+
+        if (!Strings.isNullOrEmpty(command.getPassword())) {
+            String encodedNewPassword = org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes()));
+            if (!encodedNewPassword.equals(nuageVspHost.getDetails().get("cmsuserpass"))) {
+                paramsTobeUpdated.put("cmsuserpass", org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes())));
+            }
+        }
+
+        if (command.getPort() != null &&
+                command.getPort() != Integer.parseInt(nuageVspHost.getDetails().get("port"))) {
+            paramsTobeUpdated.put("port", String.valueOf(command.getPort()));
+        }
+
+        GetClientDefaultsCommand getClientDefaultsCmd = new GetClientDefaultsCommand();
+        GetClientDefaultsAnswer getClientDefaultsAnswer = (GetClientDefaultsAnswer) _agentMgr.easySend(nuageVspHost.getId(), getClientDefaultsCmd);
+        String apiVersion = MoreObjects.firstNonNull(command.getApiVersion(), getClientDefaultsAnswer.getCurrentApiVersion());
+        SupportedApiVersionCommand supportedApiVersionCmd = new SupportedApiVersionCommand(apiVersion);
+        Answer supportedApiVersionAnswer = _agentMgr.easySend(nuageVspHost.getId(), supportedApiVersionCmd);
+        if (!supportedApiVersionAnswer.getResult()) {
+            throw new CloudRuntimeException("Incorrect API version: Nuage plugin only supports " + getClientDefaultsAnswer.getCurrentApiVersion());
+        }
+
+        String apiRelativePath = "/nuage/api/" + command.getApiVersion();
+        if (!apiRelativePath.equals(nuageVspHost.getDetails().get("apirelativepath"))) {
+            paramsTobeUpdated.put("apirelativepath", apiRelativePath);
+            paramsTobeUpdated.put("apiversion", apiVersion);
+        }
+
+        final int retryCount = Integer.parseInt(nuageVspHost.getDetails().get("retrycount"));
+        if (command.getApiRetryCount() != null && command.getApiRetryCount() != retryCount) {
```

Comment:
Move the the ``null`` check before line 385.  ``parseInt`` will fail with an NPE if the value is null.